src/app/ues/page.tsx

```diff
@@ -0,0 +1,180 @@
+};
+
+const ueFilters = Object.freeze({
+  name: { component: UENameFilter, name: '', parameterName: 'q' }, // This one does not need a name as it will never be displayed
+  creditType: { component: UECreditTypeFilter, name: 'Type de crédits', parameterName: 'creditType' },
+  branch: { component: UEBranchFilter, name: 'Branche', parameterName: 'branch' },
+  branchOption: {
+    component: UEBranchOptionFilter,
+    name: 'Filière',
+    dependsOn: ['branch'],
+    parameterName: 'branchOption',
+  },
+  semester: { component: UESemesterFilter, name: 'Semestre', parameterName: 'semester' },
+} as const satisfies {
+  [key in keyof UEFiltersType]: DeepReadonly<UEFilter<key>>;
+});
+
+type UEFilterInstance<T extends keyof UEFiltersType = keyof UEFiltersType> = {
+  filter: keyof typeof ueFilters;
+  value: UEFiltersType[T]['value'] | null;
+  search: string | null;
+};
+
+type RealUEFilterInstance<T extends keyof UEFiltersType = keyof UEFiltersType> = {
+  [K in keyof UEFilterInstance<T>]: Exclude<UEFilterInstance<T>[K], null>;
+};
+
+export default function Page() {
+  const router = useRouter();
+  const [showAddFilterDropdown, setShowAddFilterDropdown] = useState<boolean>(false);
+  const [filters, setFilters] = useState<Array<UEFilterInstance>>([{ filter: 'name', value: null, search: null }]);
+  const [lastUpdate] = useState<{ value: number }>({ value: Date.now() });
+  const [ues, updateUEs] = useSearchUEs();
+  useEffect(() => {
+    const now = Date.now();
+    lastUpdate.value = now;
+    setTimeout(() => {
+      if (lastUpdate.value === now) {
+        updateUEs(
+          Object.fromEntries(
+            filters
+              .filter((filter): filter is RealUEFilterInstance => filter.search !== null)
+              .map((filter) => [ueFilters[filter.filter].parameterName, filter.search]),
+          ),
+        );
+      }
+    }, 1000);
+  }, [filters]);
+  const updateFilter = <T extends keyof UEFiltersType>(
+    filterIndex: number,
+    value: UEFiltersType[T]['value'],
+    newUrlPart: string,
+  ) => {
+    const newFilters = [...filters];
+    newFilters[filterIndex].value = value;
+    newFilters[filterIndex].search = newUrlPart;
+    setFilters(newFilters);
+  };
+  const addFilter = (name: keyof typeof ueFilters) => {
+    setFilters([...filters, { filter: name, value: null, search: null }]);
+  };
+  const deleteFilter = (index: number) => {
+    const newFilters = filters.toSpliced(index, 1).filter(
+      (filter) =>
+        !{
+          dependsOn: [],
+          ...ueFilters[filter.filter],
+        }.dependsOn.some((dependency) => dependency === filters[index].filter),
+    );
+    setFilters(newFilters);
+  };
+  const hasFilter = (filter: keyof typeof ueFilters) => filters.some((usedFilter) => usedFilter.filter === filter);
+  const addableFilters = Object.keys(ueFilters).filter(
+    (filter) => !hasFilter(filter) && { dependsOn: [], ...ueFilters[filter] }.dependsOn.every(hasFilter),
```

Comment:
Je trouve ça plus logique/lisible, non ?
```suggestion
    (filter) => !hasFilter(filter) && (ueFilters[filter].dependsOn ?? []).every(hasFilter),
```

Reply:
la propriété `dependsOn` existe pas forcément, donc TS te sort une erreur si tu le fais comme ça (et oui, du coup ça m'a fait arrivé sur cette écriture pas fifolle)

Reply:
Oui ok, j'ai pas super bien lu; j'avais pas vu la `)`